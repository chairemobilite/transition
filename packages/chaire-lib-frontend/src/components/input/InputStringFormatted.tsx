/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { default as InputString, InputStringProps } from './InputString';

export type InputStringFormattedProps = InputStringProps & {
    /**
     * Converts the string input to the proper value type. If the input is not
     * valid, it should return null
     *
     * @memberof InputStringFormattedProps
     */
    stringToValue: (str: string) => any | null;
    /**
     * Converts the value to a string.
     *
     * @memberof InputStringFormattedProps
     */
    valueToString: (value: any) => string;
    value?: any;
    /**
     * Since this is an uncontrolled component, but it can still be changed by
     * the parent, the key is used by the parent to tell the component to update
     * the value. New props will not have effect on the value shown unless the
     * key is changed. It is a suggested pattern for components that should be
     * uncontrolled but whose value can also be updated by parent.
     */
    key?: string;
};

type InputStringFormattedState = {
    value: any;
    strValue: string;
};

/**
 * This is a wrapper to the input string widget, handling value conversion
 * to/from string as the user types input. This is an uncontrolled component,
 * that will send value back to the caller only when the entered string
 * validates. If the entered string is empty, null is returned. New props will
 * not have effect on the value shown unless the key is changed. It is a
 * suggested pattern for components that should be uncontrolled but whose value
 * can also be updated by parent
 */
export class InputStringFormatted extends React.Component<InputStringFormattedProps, InputStringFormattedState> {
    constructor(props: InputStringFormattedProps) {
        super(props);
        this.handleChange = this.handleChange.bind(this);
        this.state = {
            strValue: props.value ? props.valueToString(props.value) : '',
            value: props.value
        };
    }

    handleChange(newValue) {
        // Handle invalid values first
        if (!newValue.valid) {
            if (this.props.onValueUpdated) {
                this.props.onValueUpdated({ value: null, valid: false });
            } else if (this.props.onValueChange) {
                this.props.onValueChange(null);
            }
            this.setState({ strValue: newValue.value });
            return;
        }

        // Try to convert the formatted value for further validation
        const result = this.props.stringToValue(newValue.value);
        if (result !== null) {
            if (this.props.onValueUpdated) {
                this.props.onValueUpdated({ value: result, valid: true });
            } else if (this.props.onValueChange) {
                this.props.onValueChange(result);
            }
            this.setState({ strValue: newValue.value, value: result });
        } else if (newValue.value === '') {
            // Send null on empty strings
            if (this.props.onValueUpdated) {
                this.props.onValueUpdated({ value: null, valid: true });
            } else if (this.props.onValueChange) {
                this.props.onValueChange(null);
            }
            this.setState({ strValue: newValue.value, value: null });
        } else {
            if (this.props.onValueUpdated) {
                this.props.onValueUpdated({ value: null, valid: false });
            }
            this.setState({ strValue: newValue.value });
        }
    }

    render(): React.ReactNode {
        const { onValueUpdated, onValueChange, stringToValue, valueToString, value, ...inputProps } = this.props;
        return <InputString onValueUpdated={this.handleChange} value={this.state.strValue} {...inputProps} />;
    }
}

export default InputStringFormatted;
