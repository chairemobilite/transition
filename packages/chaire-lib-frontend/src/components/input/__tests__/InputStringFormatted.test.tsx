/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import InputStringFormatted from '../InputStringFormatted';

const mockOnChange = jest.fn();
const testId = 'StringFormattedWidgetId';
const intToStr = (val: number): string => val.toString();
const strToInt = (str: string) => (str === (intToStr(Number.parseInt(str))) ? Number.parseInt(str) : null);
const originalIntValue = 10;

beforeEach(() => {
    mockOnChange.mockClear();
});

test('Default props', () => {
    const { container } = render(<InputStringFormatted
        id = {testId}
        onValueUpdated = {mockOnChange}
        stringToValue = {strToInt}
        valueToString = {intToStr}
        key = {testId}
    />);
    expect(container).toMatchSnapshot();
});

test('All props', () => {
    const { container } = render(<InputStringFormatted
        id = {testId}
        onValueUpdated = {mockOnChange}
        value = "test"
        maxLength = {100}
        disabled = {false}
        autocompleteChoices = {[]}
        stringToValue = {strToInt}
        valueToString = {intToStr}
        key = {testId}
    />);
    expect(container).toMatchSnapshot();
});

test('Disabled', () => {
    const { container } = render(<InputStringFormatted
        id = {testId}
        onValueUpdated = {mockOnChange}
        disabled = {true}
        stringToValue = {strToInt}
        valueToString = {intToStr}
        key = {testId}
    />);
    expect(container).toMatchSnapshot();
});

test('Text input change with deprecated onValueChange', () => {
    const { container } = render(<InputStringFormatted
        id = {testId}
        onValueChange = {mockOnChange}
        value = {originalIntValue}
        maxLength = {100}
        stringToValue = {strToInt}
        valueToString = {intToStr}
        key = {testId}
    />);
    // Validate initial values
    const inputElement = container.querySelector(`#${testId}`) as HTMLInputElement;
    expect(inputElement.value).toBe(originalIntValue.toString());

    // Change the value manually to a valid value
    fireEvent.change(inputElement, { target: { value: (originalIntValue + 1).toString() } });
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith(originalIntValue + 1);
    expect(inputElement.value).toBe((originalIntValue + 1).toString());

    // Manually change the value to something that doesn't pass the string validation
    fireEvent.change(inputElement, { target: { value: (originalIntValue + 1).toString() + 'a' } });
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(inputElement.value).toBe((originalIntValue + 1).toString() + 'a');

    // Set the value to empty string
    fireEvent.change(inputElement, { target: { value: '' } });
    expect(mockOnChange).toHaveBeenCalledTimes(2);
    expect(mockOnChange).toHaveBeenCalledWith(null);
});

test('Text input change', () => {
    const { container } = render(<InputStringFormatted
        id = {testId}
        onValueUpdated = {mockOnChange}
        value = {originalIntValue}
        maxLength = {100}
        stringToValue = {strToInt}
        valueToString = {intToStr}
        key = {testId}
    />);
    // Validate initial values
    const inputElement = container.querySelector(`#${testId}`) as HTMLInputElement;
    expect(inputElement.value).toBe(originalIntValue.toString());

    // Change the value manually to a valid value
    fireEvent.change(inputElement, { target: { value: (originalIntValue + 1).toString() } });
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith({ value: originalIntValue + 1, valid: true });
    expect(inputElement.value).toBe((originalIntValue + 1).toString());

    // Manually change the value to something that doesn't pass the string validation
    fireEvent.change(inputElement, { target: { value: (originalIntValue + 1).toString() + 'a' } });
    expect(mockOnChange).toHaveBeenCalledTimes(2);
    expect(mockOnChange).toHaveBeenLastCalledWith({ value: null, valid: false });
    expect(inputElement.value).toBe((originalIntValue + 1).toString() + 'a');

    // Set the value to empty string
    fireEvent.change(inputElement, { target: { value: '' } });
    expect(mockOnChange).toHaveBeenCalledTimes(3);
    expect(mockOnChange).toHaveBeenLastCalledWith({ value: null, valid: true });
});

test('Test with type and pattern', () => {
    /* TODO tahini: testing with those props does not seem to work properly */
});

