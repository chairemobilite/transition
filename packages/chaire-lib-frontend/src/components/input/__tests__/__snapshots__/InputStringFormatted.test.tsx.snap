// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All props 1`] = `
<div>
  <input
    autocomplete="none"
    class="apptr__form-input-string apptr__input apptr__input-string"
    id="StringFormattedWidgetId"
    maxlength="100"
    name="StringFormattedWidgetId"
    type="text"
    value="test"
  />
</div>
`;

exports[`Default props 1`] = `
<div>
  <input
    autocomplete="none"
    class="apptr__form-input-string apptr__input apptr__input-string"
    id="StringFormattedWidgetId"
    maxlength="255"
    name="StringFormattedWidgetId"
    type="text"
    value=""
  />
</div>
`;

exports[`Disabled 1`] = `
<div>
  <input
    autocomplete="none"
    class="apptr__form-input-string apptr__input apptr__input-string"
    disabled=""
    id="StringFormattedWidgetId"
    maxlength="255"
    name="StringFormattedWidgetId"
    type="text"
    value=""
  />
</div>
`;
